internal/provider/resource_workspace_run.go

```diff
@@ -0,0 +1,386 @@
+	req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	// Retrieve values from workspace run.
+	var workspaceRun WorkspaceRunModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &workspaceRun)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	created := t.doApplyOrDestroyRun(ctx, workspaceRun, false, resp.Diagnostics)
+
+	// Update the plan with the computed attribute values.
+	t.copyWorkspaceRun(created, &workspaceRun)
+
+	// Set the response state to the fully-populated plan, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, workspaceRun)...)
+}
+
+func (t *workspaceRunResource) Read(ctx context.Context,
+	req resource.ReadRequest, resp *resource.ReadResponse) {
+
+	// Get the current state.
+	var state WorkspaceRunModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get latest run on the target workspace.
+	toSortBy := sdktypes.RunSortableFieldUpdatedAtDesc // variable needed because can't take address of constant
+	one := int32(1)                                    // ditto
+	gotRuns, err := t.client.Run.GetRuns(ctx, &sdktypes.GetRunsInput{
+		Sort:              &toSortBy,
+		PaginationOptions: &sdktypes.PaginationOptions{Limit: &one},
+		Filter: &sdktypes.RunFilter{
+			WorkspacePath: ptr.String(state.WorkspacePath.ValueString()),
+		},
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to get runs for target workspace", err.Error())
+		return
+	}
+	if gotRuns.Runs == nil {
+		resp.Diagnostics.AddError("GetRuns on target workspace returned nil.", "")
+		return
+	}
+	if len(gotRuns.Runs) == 0 {
+		resp.Diagnostics.AddError("GetRuns on target workspace returned empty", "")
+		return
+	}
+	latestRun := gotRuns.Runs[0]
+
+	// Make sure the module source and module version are not nil.
+	if latestRun.ModuleSource == nil {
+		resp.Diagnostics.AddError("No module source available", fmt.Sprintf("for workspace %s", latestRun.WorkspacePath))
+		return
+	}
+	if latestRun.ModuleVersion == nil {
+		resp.Diagnostics.AddError("No module version available", fmt.Sprintf("for workspace %s", latestRun.WorkspacePath))
+		return
+	}
+
+	// Update the state with the computed attribute values.
+	t.copyWorkspaceRun(&WorkspaceRunModel{
+		WorkspacePath: state.WorkspacePath,
+		ModuleSource:  types.StringValue(*latestRun.ModuleSource),
+		ModuleVersion: types.StringValue(*latestRun.ModuleVersion),
+		Variables:     state.Variables,
+	}, &state)
+
+	// Set the refreshed state, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (t *workspaceRunResource) Update(ctx context.Context,
+	req resource.UpdateRequest, resp *resource.UpdateResponse) {
+
+	// Retrieve values from plan.
+	var plan WorkspaceRunModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// FIXME: See other review items to set this correctly.
+	isDestroyRun := false
```

Comment:
The update and delete should only run if the currently deployed module/version doesn't match the desired module/version.